File: frontend/src/lib/types/graph.ts
export interface GraphNode {
  id:         string;
  type:       string;
  label:      string;
  properties: Record<string, unknown>;
}

export interface GraphEdge {
  source_id:    string;
  source_type:  string;
  target_id:    string;
  target_type:  string;
  relationship: string;
  weight:       number;
  properties:   Record<string, unknown>;
}

export interface GraphNeighborhoodResponse {
  root_type:   string;
  root_id:     string;
  depth:       number;
  nodes:       GraphNode[];
  edges:       GraphEdge[];
  total_nodes: number;
  total_edges: number;
}

export interface CentralityNode {
  node_type:       string;
  node_id:         string;
  label:           string;
  degree:          number;
  in_degree:       number;
  out_degree:      number;
  weighted_degree: number;
}

export interface GraphStatsResponse {
  total_nodes:           number;
  total_edges:           number;
  nodes_by_type:         Record<string, number>;
  edges_by_relationship: Record<string, number>;
  top_central_nodes:     CentralityNode[];
}

File: api/schemas/graph.py
"""Typed API contracts for graph traversal and visualization endpoints."""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class GraphNodeSchema(BaseModel):
    id:         str
    type:       str     # covered_entity | pharmacy | provider | ndc | finding | case
    label:      str
    properties: dict[str, Any] = Field(default_factory=dict)


class GraphEdgeSchema(BaseModel):
    source_id:    str
    source_type:  str
    target_id:    str
    target_type:  str
    relationship: str
    weight:       float = 1.0
    properties:   dict[str, Any] = Field(default_factory=dict)


class GraphNeighborhoodResponse(BaseModel):
    root_type:   str
    root_id:     str
    depth:       int
    nodes:       list[GraphNodeSchema] = Field(default_factory=list)
    edges:       list[GraphEdgeSchema] = Field(default_factory=list)
    total_nodes: int = 0
    total_edges: int = 0


class CentralityNode(BaseModel):
    node_type:       str
    node_id:         str
    label:           str
    degree:          int
    in_degree:       int
    out_degree:      int
    weighted_degree: float


class GraphStatsResponse(BaseModel):
    total_nodes:          int
    total_edges:          int
    nodes_by_type:        dict[str, int]
    edges_by_relationship: dict[str, int]
    top_central_nodes:    list[CentralityNode] = Field(default_factory=list)
